### PR TITLE
Ignore nested projects of same name as root

### PR DESCRIPTION
**Pull request by seth**
This is a fix for #1779

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,6 +37,7 @@ Cargo requires the following tools and packages to build:
 * `cmake`
 * `pkg-config`
 * OpenSSL headers (`libssl-dev` package on ubuntu)
+* libgit2
```
Comment (alexcrichton) [inline README.md]: Currently libgit2 is built from source if it's not found on the system, so this shouldn't be necessary as a build prerequisite.

**Comment by alexcrichton** [comment]
Could you add a test for this as well?

File: README.md (hunk 1 above)
Comment (seth) [inline README.md]: Hmm, that wasn't my experience building on OS X. Perhaps this should be an OS X specific note? Or perhaps I've missed something.

Comment (alexcrichton) [inline README.md]: What's the error message/build logs of not having libgit2 installed?

Comment (seth) [inline README.md]: Here's what I was seeing originally:
https://gist.github.com/seth/ef06472d873885075c4a

However, I believe I tracked this down to having accidentally had an alternate pkg-config binary on my PATH.

I just tried `brew uninstall libgit2 && make clean && make` and it built fine. So I have a feeling the funkiness of my environment was the issue. Will remove this note.

**Comment by seth** [comment]
I'd be happy to add a test. It would be super helpful to have a pointer or two on where that'd go and at what level you would like to see tested.

Looks like `test_cargo_package.rs` might be the place to add something?

**Comment by seth** [comment]
I think I have a test, will push shortly.

**Comment by seth** [comment]
@alexcrichton test added. It failed for me when I removed my patch.

File: src/cargo/ops/cargo_read_manifest.rs [hunk 2]
```diff
@@ -112,7 +112,13 @@ fn read_nested_packages(path: &Path,
     let manifest = try!(find_project_manifest_exact(path, "Cargo.toml"));
 
     let (pkg, nested) = try!(read_package(&manifest, source_id, config));
-    all_packages.insert(pkg.package_id().clone(), pkg);
+    let pkg_id = pkg.package_id().clone();
+    if !all_packages.contains_key(&pkg_id) {
+        all_packages.insert(pkg_id, pkg);
+    } else {
+        warn!("skipping nested package `{}` found at `{}`",
+              &pkg.name(), &path.to_string_lossy());
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_read_manifest.rs]: You should be able to omit the leading `&` characters here.

Can you also switch this to a `info!` from `warn!`?

**Comment by alexcrichton** [comment]
I'm a little worried about this in the sense that it just blindly takes the first package with a particular name (if there are multiple with the same names in a repo) and we don't have any real guarantee about what order the repo is traversed in. That being said, I think that the semantics are of having multiple packages of the same name in a repo aren't super well defined today, so I think it's fine to just take the first one.

**Comment by seth** [comment]
Updated with info! and less & :)

I can see what you mean about the potential ambiguity. For the case I ran into, I think the top-level package being read first is likely to stay consistent?

**Comment by alexcrichton** [comment]
@bors: r+ 8790c820b690c27282d564bd399a9fc31c7ff3a7

Thanks!

> I think the top-level package being read first is likely to stay consistent

Yeah this is definitely true.

**Pull request merged**